File: .claude/skills/task-spec/references/examples/consume-task-spec.py
#!/usr/bin/env python3
"""Minimal Task-Spec v2.1.1 consumer.

Parses a T-*.md file, validates the YAML frontmatter against the published
JSON Schema, extracts the validation_card block (agent_contract + retry_policy
+ success_criteria), enumerates eval_N() bash blocks, and prints a structured
TaskSpec dataclass.

Runtime: Python 3.8+. Third-party deps: PyYAML (always), jsonschema (optional;
skipped with a warning if not installed). Schemas are resolved relative to
this file: ../schemas/task-spec-frontmatter.schema.json.

Usage:
    python3 consume-task-spec.py <path/to/T-*.md>

Exit codes:
    0 — parsed and validated cleanly
    1 — file missing / unreadable / frontmatter malformed
    2 — schema validation failed
"""
from __future__ import annotations

import json
import re
import sys
from dataclasses import dataclass, field, asdict
from pathlib import Path
from typing import Any, Dict, List, Optional

try:
    import yaml
except ImportError:
    print("ERROR: PyYAML is required. Install with: pip install pyyaml", file=sys.stderr)
    sys.exit(1)

SCHEMAS_DIR = Path(__file__).resolve().parent.parent / "schemas"
FRONTMATTER_SCHEMA = SCHEMAS_DIR / "task-spec-frontmatter.schema.json"
CONTRACT_SCHEMA = SCHEMAS_DIR / "agent-contract.schema.json"

EVAL_FN_RE = re.compile(r"^(eval_\d+)\s*\(\)\s*\{", re.MULTILINE)


class _StrTimestampLoader(yaml.SafeLoader):
    """SafeLoader variant that keeps ISO timestamps as strings.

    PyYAML's default SafeLoader converts ``2026-06-02T00:00:00Z`` to a
    ``datetime`` object, which breaks JSON Schema string validation.
    """


_StrTimestampLoader.add_constructor(
    "tag:yaml.org,2002:timestamp",
    lambda loader, node: loader.construct_scalar(node),
)


def _yaml_load(text: str) -> Any:
    return yaml.load(text, Loader=_StrTimestampLoader)


@dataclass
class TaskSpec:
    """Parsed view of a Task-Spec v2.1.1 file."""

    path: str
    id: str
    title: str
    status: str
    format_version: Any
    execution_backend: str
    signed_off: bool
    frontmatter: Dict[str, Any]
    validation_card: Dict[str, Any] = field(default_factory=dict)
    eval_ids: List[str] = field(default_factory=list)


def _split_frontmatter(text: str) -> (Dict[str, Any], str):
    if not text.startswith("---"):
        raise ValueError("file does not start with YAML frontmatter '---'")
    parts = text.split("---\n", 2)
    if len(parts) < 3:
        raise ValueError("frontmatter not terminated by a closing '---' line")
    return _yaml_load(parts[1]) or {}, parts[2]


def _extract_validation_card(body: str) -> Dict[str, Any]:
    marker = "## Validation Card"
    idx = body.find(marker)
    if idx == -1:
        return {}
    after = body[idx + len(marker):]
    fence = after.find("```yaml")
    if fence == -1:
        return {}
    after = after[fence + len("```yaml"):]
    end = after.find("```")
    if end == -1:
        return {}
    return _yaml_load(after[:end]) or {}


def _extract_eval_ids(body: str) -> List[str]:
    marker = "## Success Criteria"
    idx = body.find(marker)
    if idx == -1:
        return []
    after = body[idx + len(marker):]
    end = after.find("\n## ")
    block = after if end == -1 else after[:end]
    return [m.group(1) for m in EVAL_FN_RE.finditer(block)]


def _validate(instance: Dict[str, Any], schema_path: Path, label: str) -> None:
    try:
        from jsonschema import Draft202012Validator
    except ImportError:
        print(f"WARN: jsonschema not installed; skipping {label} validation", file=sys.stderr)
        return
    with schema_path.open() as f:
        schema = json.load(f)
    errors = sorted(Draft202012Validator(schema).iter_errors(instance), key=lambda e: e.path)
    if errors:
        for e in errors:
            print(f"FAIL {label}: {'/'.join(str(p) for p in e.path) or '(root)'} — {e.message}", file=sys.stderr)
        sys.exit(2)


def parse(path: Path) -> TaskSpec:
    text = path.read_text(encoding="utf-8")
    fm, body = _split_frontmatter(text)
    card = _extract_validation_card(body)
    eval_ids = _extract_eval_ids(body)
    _validate(fm, FRONTMATTER_SCHEMA, "frontmatter")
    if card:
        _validate(card, CONTRACT_SCHEMA, "validation_card")
    return TaskSpec(
        path=str(path),
        id=str(fm.get("id", "")),
        title=str(fm.get("title", "")),
        status=str(fm.get("status", "")),
        format_version=fm.get("format_version"),
        execution_backend=str(fm.get("execution_backend", "")),
        signed_off=bool(fm.get("signed_off", False)),
        frontmatter=fm,
        validation_card=card,
        eval_ids=eval_ids,
    )


def main(argv: Optional[List[str]] = None) -> int:
    args = argv if argv is not None else sys.argv[1:]
    if not args:
        print("Usage: consume-task-spec.py <path/to/T-*.md>", file=sys.stderr)
        return 1
    target = Path(args[0])
    if not target.is_file():
        print(f"ERROR: {target} not found", file=sys.stderr)
        return 1
    spec = parse(target)
    summary = asdict(spec)
    summary["frontmatter"] = {k: spec.frontmatter.get(k) for k in ("id", "status", "format_version", "execution_backend", "signed_off")}
    print(json.dumps(summary, indent=2, default=str))
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: .claude/skills/task-spec/references/examples/consume-task-spec.ts
#!/usr/bin/env -S npx ts-node
/**
 * Minimal Task-Spec v2.1.1 consumer (TypeScript).
 *
 * Parses a T-*.md file, validates the YAML frontmatter against the published
 * JSON Schema, extracts the validation_card block (agent_contract +
 * retry_policy + success_criteria), enumerates eval_N() bash blocks, and
 * prints a structured TaskSpec record.
 *
 * Runtime: Node 18+. Third-party deps: `yaml` (parser) and `ajv` (validator),
 * pinned in references/examples/package.json. Schemas are resolved relative
 * to this file: ../schemas/task-spec-frontmatter.schema.json.
 *
 * Usage:
 *   npx ts-node consume-task-spec.ts <path/to/T-*.md>
 *
 * Exit codes:
 *   0 — parsed and validated cleanly
 *   1 — file missing / unreadable / frontmatter malformed
 *   2 — schema validation failed
 */
import { readFileSync, existsSync } from "node:fs";
import { resolve, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import YAML from "yaml";
import Ajv2020 from "ajv/dist/2020.js";

const __filename =
  typeof __dirname !== "undefined"
    ? `${__dirname}/consume-task-spec.ts`
    : fileURLToPath(import.meta.url);
const HERE = dirname(__filename);
const SCHEMAS_DIR = resolve(HERE, "..", "schemas");
const FRONTMATTER_SCHEMA = resolve(SCHEMAS_DIR, "task-spec-frontmatter.schema.json");
const CONTRACT_SCHEMA = resolve(SCHEMAS_DIR, "agent-contract.schema.json");

const EVAL_FN_RE = /^(eval_\d+)\s*\(\)\s*\{/gm;

interface TaskSpec {
  path: string;
  id: string;
  title: string;
  status: string;
  format_version: unknown;
  execution_backend: string;
  signed_off: boolean;
  frontmatter: Record<string, unknown>;
  validation_card: Record<string, unknown>;
  eval_ids: string[];
}

function splitFrontmatter(text: string): { frontmatter: Record<string, unknown>; body: string } {
  if (!text.startsWith("---")) {
    throw new Error("file does not start with YAML frontmatter '---'");
  }
  const parts = text.split("---\n");
  if (parts.length < 3) {
    throw new Error("frontmatter not terminated by a closing '---' line");
  }
  const yamlText = parts[1];
  const body = parts.slice(2).join("---\n");
  const fm = (YAML.parse(yamlText) ?? {}) as Record<string, unknown>;
  return { frontmatter: fm, body };
}

function extractValidationCard(body: string): Record<string, unknown> {
  const marker = "## Validation Card";
  const idx = body.indexOf(marker);
  if (idx === -1) return {};
  const after = body.slice(idx + marker.length);
  const fenceStart = after.indexOf("```yaml");
  if (fenceStart === -1) return {};
  const fromYaml = after.slice(fenceStart + "```yaml".length);
  const fenceEnd = fromYaml.indexOf("```");
  if (fenceEnd === -1) return {};
  return (YAML.parse(fromYaml.slice(0, fenceEnd)) ?? {}) as Record<string, unknown>;
}

function extractEvalIds(body: string): string[] {
  const marker = "## Success Criteria";
  const idx = body.indexOf(marker);
  if (idx === -1) return [];
  const after = body.slice(idx + marker.length);
  const end = after.indexOf("\n## ");
  const block = end === -1 ? after : after.slice(0, end);
  const ids: string[] = [];
  let match: RegExpExecArray | null;
  while ((match = EVAL_FN_RE.exec(block)) !== null) {
    ids.push(match[1]);
  }
  return ids;
}

function validate(
  instance: unknown,
  schemaPath: string,
  label: string,
  ajv: Ajv2020,
): void {
  if (!existsSync(schemaPath)) {
    process.stderr.write(`WARN: ${label} schema missing at ${schemaPath}\n`);
    return;
  }
  const schema = JSON.parse(readFileSync(schemaPath, "utf8"));
  const validateFn = ajv.compile(schema);
  if (!validateFn(instance)) {
    for (const err of validateFn.errors ?? []) {
      process.stderr.write(`FAIL ${label}: ${err.instancePath || "(root)"} — ${err.message}\n`);
    }
    process.exit(2);
  }
}

function parse(path: string): TaskSpec {
  const text = readFileSync(path, "utf8");
  const { frontmatter, body } = splitFrontmatter(text);
  const card = extractValidationCard(body);
  const evalIds = extractEvalIds(body);
  const ajv = new Ajv2020({ strict: false, allErrors: true });
  validate(frontmatter, FRONTMATTER_SCHEMA, "frontmatter", ajv);
  if (Object.keys(card).length > 0) {
    validate(card, CONTRACT_SCHEMA, "validation_card", ajv);
  }
  return {
    path,
    id: String(frontmatter["id"] ?? ""),
    title: String(frontmatter["title"] ?? ""),
    status: String(frontmatter["status"] ?? ""),
    format_version: frontmatter["format_version"],
    execution_backend: String(frontmatter["execution_backend"] ?? ""),
    signed_off: Boolean(frontmatter["signed_off"] ?? false),
    frontmatter,
    validation_card: card,
    eval_ids: evalIds,
  };
}

function main(argv: string[]): number {
  if (argv.length === 0) {
    process.stderr.write("Usage: consume-task-spec.ts <path/to/T-*.md>\n");
    return 1;
  }
  const target = resolve(argv[0]);
  if (!existsSync(target)) {
    process.stderr.write(`ERROR: ${target} not found\n`);
    return 1;
  }
  const spec = parse(target);
  const summary = {
    ...spec,
    frontmatter: {
      id: spec.frontmatter["id"],
      status: spec.frontmatter["status"],
      format_version: spec.frontmatter["format_version"],
      execution_backend: spec.frontmatter["execution_backend"],
      signed_off: spec.frontmatter["signed_off"],
    },
  };
  process.stdout.write(JSON.stringify(summary, null, 2) + "\n");
  return 0;
}

process.exit(main(process.argv.slice(2)));
